// discordeno@18.0.1/script/helpers/channels/editChannelPositions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.editChannelPositions = exports.swapChannels = void 0;
exports.swapChannels = editChannelPositions;
/**
 * Edits the positions of a set of channels in a guild.
 *
 * @param bot - The bot instance to use to make the request.
 * @param guildId - The ID of the guild in which to edit the positions of the channels.
 * @param channelPositions - A set of objects defining the updated positions of the channels.
 *
 * @remarks
 * Requires the `MANAGE_CHANNELS` permission.
 *
 * Fires a _Channel Update_ gateway event for every channel impacted in this change.
 *
 * @see {@link https://discord.com/developers/docs/resources/guild#modify-guild-channel-positions}
 */
async function editChannelPositions(bot, guildId, channelPositions) {
    if (!channelPositions.length) {
        throw new Error("You must provide at least one channels to be moved.");
    }
    return await bot.rest.runMethod(bot.rest, "PATCH", bot.constants.routes.GUILD_CHANNELS(guildId), channelPositions.map((channelPosition) => ({
        id: channelPosition.id,
        position: channelPosition.position,
        lock_positions: channelPosition.lockPositions,
        parent_id: channelPosition.parentId,
    })));
}
exports.editChannelPositions = editChannelPositions;
