// discordeno/script/transformers/reverse/createApplicationCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformCreateApplicationCommandToDiscordCreateApplicationCommand = void 0;
const mod_js_1 = require("../../types/mod.js");
function transformCreateApplicationCommandToDiscordCreateApplicationCommand(bot, payload) {
    if ((0, mod_js_1.isContextApplicationCommand)(payload)) {
        return {
            name: payload.name,
            name_localizations: payload.nameLocalizations,
            description: "",
            description_localizations: {},
            type: payload.type,
            default_member_permissions: payload.defaultMemberPermissions
                ? bot.utils.calculateBits(payload.defaultMemberPermissions)
                : null,
            dm_permission: payload.dmPermission,
        };
    }
    return {
        name: payload.name,
        name_localizations: payload.nameLocalizations,
        description: payload.description,
        description_localizations: payload.descriptionLocalizations,
        type: payload.type,
        options: payload.options?.map((option) => bot.transformers.reverse.applicationCommandOption(bot, option)),
        default_member_permissions: payload.defaultMemberPermissions
            ? bot.utils.calculateBits(payload.defaultMemberPermissions)
            : null,
        dm_permission: payload.dmPermission,
    };
}
exports.transformCreateApplicationCommandToDiscordCreateApplicationCommand = transformCreateApplicationCommandToDiscordCreateApplicationCommand;
