// discordeno@18.0.1/script/helpers/messages/editMessage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.editMessage = void 0;
/**
 * Edits a message.
 *
 * @param bot - The bot instance to use to make the request.
 * @param channelId - The ID of the channel to edit the message in.
 * @param messageId - The IDs of the message to edit.
 * @param options - The parameters for the edit of the message.
 * @returns An instance of the edited {@link Message}.
 *
 * @remarks
 * If editing another user's message:
 * - Requires the `MANAGE_MESSAGES` permission.
 * - Only the {@link EditMessage.flags | flags} property of the {@link options} object parameter can be edited.
 *
 * Fires a _Message Update_ gateway event.
 *
 * @see {@link https://discord.com/developers/docs/resources/channel#edit-message}
 */
async function editMessage(bot, channelId, messageId, options) {
    const result = await bot.rest.runMethod(bot.rest, "PATCH", bot.constants.routes.CHANNEL_MESSAGE(channelId, messageId), {
        content: options.content,
        embeds: options.embeds?.map((embed) => bot.transformers.reverse.embed(bot, embed)),
        allowed_mentions: options.allowedMentions
            ? bot.transformers.reverse.allowedMentions(bot, options.allowedMentions)
            : undefined,
        attachments: options.attachments?.map((attachment) => bot.transformers.reverse.attachment(bot, attachment)),
        file: options.file,
        components: options.components?.map((component) => bot.transformers.reverse.component(bot, component)),
    });
    return bot.transformers.message(bot, result);
}
exports.editMessage = editMessage;
