// discordeno@18.0.1/script/helpers/messages/sendMessage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sendMessage = void 0;
const shared_js_1 = require("../../types/shared.js");
/**
 * Sends a message to a channel.
 *
 * @param bot - The bot instance to use to make the request.
 * @param channelId - The ID of the channel to send the message in.
 * @param options - The parameters for the creation of the message.
 * @returns An instance of the created {@link Message}.
 *
 * @remarks
 * Requires that the bot user be able to see the contents of the channel the message is to be sent in.
 *
 * If sending a message to a guild channel:
 * - Requires the `SEND_MESSAGES` permission.
 *
 * If sending a TTS message:
 * - Requires the `SEND_TTS_MESSAGES` permission.
 *
 * If sending a message as a reply to another message:
 * - Requires the `READ_MESSAGE_HISTORY` permission.
 * - The message being replied to cannot be a system message.
 *
 * ⚠️ The maximum size of a request (accounting for any attachments and message content) for bot users is _8 MiB_.
 *
 * Fires a _Message Create_ gateway event.
 *
 * @see {@link https://discord.com/developers/docs/resources/channel#create-message}
 */
async function sendMessage(bot, channelId, options) {
    const result = await bot.rest.runMethod(bot.rest, "POST", bot.constants.routes.CHANNEL_MESSAGES(channelId), {
        content: options.content,
        nonce: options.nonce,
        tts: options.tts,
        embeds: options.embeds?.map((embed) => bot.transformers.reverse.embed(bot, embed)),
        allowed_mentions: options.allowedMentions
            ? {
                parse: options.allowedMentions?.parse,
                roles: options.allowedMentions?.roles?.map((id) => id.toString()),
                users: options.allowedMentions?.users?.map((id) => id.toString()),
                replied_user: options.allowedMentions?.repliedUser,
            }
            : undefined,
        file: options.file,
        components: options.components?.map((component) => ({
            type: component.type,
            components: component.components.map((subComponent) => {
                if (subComponent.type === shared_js_1.MessageComponentTypes.InputText) {
                    return {
                        type: subComponent.type,
                        style: subComponent.style,
                        custom_id: subComponent.customId,
                        label: subComponent.label,
                        placeholder: subComponent.placeholder,
                        min_length: subComponent.minLength ?? subComponent.required === false ? 0 : subComponent.minLength,
                        max_length: subComponent.maxLength,
                    };
                }
                if (subComponent.type === shared_js_1.MessageComponentTypes.SelectMenu) {
                    return {
                        type: subComponent.type,
                        custom_id: subComponent.customId,
                        placeholder: subComponent.placeholder,
                        min_values: subComponent.minValues,
                        max_values: subComponent.maxValues,
                        disabled: "disabled" in subComponent ? subComponent.disabled : undefined,
                        options: subComponent.options.map((option) => ({
                            label: option.label,
                            value: option.value,
                            description: option.description,
                            emoji: option.emoji
                                ? {
                                    id: option.emoji.id?.toString(),
                                    name: option.emoji.name,
                                    animated: option.emoji.animated,
                                }
                                : undefined,
                            default: option.default,
                        })),
                    };
                }
                if (subComponent.type === shared_js_1.MessageComponentTypes.SelectMenuChannels ||
                    subComponent.type === shared_js_1.MessageComponentTypes.SelectMenuRoles ||
                    subComponent.type === shared_js_1.MessageComponentTypes.SelectMenuUsers ||
                    subComponent.type === shared_js_1.MessageComponentTypes.SelectMenuUsersAndRoles) {
                    return {
                        type: subComponent.type,
                        custom_id: subComponent.customId,
                        placeholder: subComponent.placeholder,
                        min_values: subComponent.minValues,
                        max_values: subComponent.maxValues,
                        disabled: "disabled" in subComponent ? subComponent.disabled : undefined,
                    };
                }
                return {
                    type: subComponent.type,
                    custom_id: subComponent.customId,
                    label: subComponent.label,
                    style: subComponent.style,
                    emoji: "emoji" in subComponent && subComponent.emoji
                        ? {
                            id: subComponent.emoji.id?.toString(),
                            name: subComponent.emoji.name,
                            animated: subComponent.emoji.animated,
                        }
                        : undefined,
                    url: "url" in subComponent ? subComponent.url : undefined,
                    disabled: "disabled" in subComponent ? subComponent.disabled : undefined,
                };
            }),
        })),
        ...(options.messageReference?.messageId
            ? {
                message_reference: {
                    message_id: options.messageReference.messageId.toString(),
                    channel_id: options.messageReference.channelId?.toString(),
                    guild_id: options.messageReference.guildId?.toString(),
                    fail_if_not_exists: options.messageReference.failIfNotExists === true,
                },
            }
            : {}),
        sticker_ids: options.stickerIds?.map((sticker) => sticker.toString()),
    });
    return bot.transformers.message(bot, result);
}
exports.sendMessage = sendMessage;
