// discordeno@18.0.1/script/helpers/members/editMember.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.editMember = void 0;
/**
 * Edits a member's properties.
 *
 * @param bot - The bot instance to use to make the request.
 * @param guildId - The ID of the guild to edit the member of.
 * @param userId - The user ID of the member to edit.
 * @param options - The parameters for the edit of the user.
 *
 * @remarks
 * This endpoint requires various permissions depending on what is edited about the member.
 * To find out the required permission to enact a change, read the documentation of this endpoint's {@link ModifyGuildMember | parameters}.
 *
 * Fires a _Guild Member Update_ gateway event.
 *
 * @see {@link https://discord.com/developers/docs/resources/guild#modify-guild-member}
 */
async function editMember(bot, guildId, userId, options) {
    const result = await bot.rest.runMethod(bot.rest, "PATCH", bot.constants.routes.GUILD_MEMBER(guildId, userId), {
        nick: options.nick,
        roles: options.roles?.map((id) => id.toString()),
        mute: options.mute,
        deaf: options.deaf,
        channel_id: options.channelId?.toString(),
        communication_disabled_until: options.communicationDisabledUntil
            ? new Date(options.communicationDisabledUntil).toISOString()
            : options.communicationDisabledUntil,
    });
    return bot.transformers.member(bot, result, bot.transformers.snowflake(guildId), bot.transformers.snowflake(userId));
}
exports.editMember = editMember;
