// @pnpm/registry-mock@3.8.0/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.REGISTRY_MOCK_PORT = exports.getIntegrity = exports.addUser = exports.addDistTag = exports.start = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const execa_1 = __importDefault(require("execa"));
const locations = __importStar(require("./lib/locations"));
const addDistTag_1 = __importDefault(require("./lib/addDistTag"));
const addUser_1 = __importDefault(require("./lib/addUser"));
const REGISTRY_MOCK_PORT = locations.REGISTRY_MOCK_PORT;
exports.REGISTRY_MOCK_PORT = REGISTRY_MOCK_PORT;
function default_1() {
    const verdaccioBin = require.resolve('verdaccio/bin/verdaccio');
    return execa_1.default.sync('node', [
        verdaccioBin,
        '--config',
        locations.configPath(),
        '--listen',
        REGISTRY_MOCK_PORT
    ], { stdio: 'inherit' });
}
exports.default = default_1;
function start(opts) {
    const verdaccioBin = require.resolve('verdaccio/bin/verdaccio');
    return (0, execa_1.default)('node', [
        verdaccioBin,
        '--config',
        locations.configPath(),
        '--listen',
        REGISTRY_MOCK_PORT
    ], opts);
}
exports.start = start;
exports.addDistTag = (0, addDistTag_1.default)(REGISTRY_MOCK_PORT);
exports.addUser = (0, addUser_1.default)(REGISTRY_MOCK_PORT);
const getIntegrity = (pkgName, pkgVersion) => {
    return JSON.parse(fs_1.default.readFileSync(path_1.default.join(locations.storage(), pkgName, 'package.json'), 'utf8')).versions[pkgVersion].dist.integrity;
};
exports.getIntegrity = getIntegrity;
