// .ignored/@thereallonewolf/amethystframework/script/src/inhibators/mod.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.inhibitors = void 0;
// deno-lint-ignore-file
const dntShim = __importStar(require("../../_dnt.shims.js"));
const mod_js_1 = require("../../mod.js");
const errors_js_1 = require("../interfaces/errors.js");
const AmethystCollection_js_1 = require("../utils/AmethystCollection.js");
exports.inhibitors = new AmethystCollection_js_1.AmethystCollection();
const membersInCooldown = new Map();
exports.inhibitors.set("cooldown", async (bot, command, options) => {
    const commandCooldown = command.cooldown || bot.defaultCooldown;
    if (!commandCooldown ||
        (options?.author.id && bot.ignoreCooldown?.includes(options?.author.id)))
        return true;
    const key = `${options.author.id}-${command.name}`;
    const cooldown = membersInCooldown.get(key);
    if (cooldown) {
        if (cooldown.used >= (commandCooldown.allowedUses || 1)) {
            const now = Date.now();
            if (cooldown.timestamp > now) {
                return {
                    type: errors_js_1.ErrorEnums.COOLDOWN,
                    value: {
                        expiresAt: Date.now() + commandCooldown.seconds * 1000,
                        executedAt: Date.now(),
                    },
                };
            }
            else {
                cooldown.used = 0;
            }
        }
        membersInCooldown.set(key, {
            used: cooldown.used + 1,
            timestamp: Date.now() + commandCooldown.seconds * 1000,
        });
        return {
            type: errors_js_1.ErrorEnums.COOLDOWN,
            value: {
                expiresAt: Date.now() + commandCooldown.seconds * 1000,
                executedAt: Date.now(),
            },
        };
    }
    membersInCooldown.set(key, {
        used: 1,
        timestamp: Date.now() + commandCooldown.seconds * 1000,
    });
    return true;
});
dntShim.setInterval(() => {
    const now = Date.now();
    membersInCooldown.forEach((cooldown, key) => {
        if (cooldown.timestamp > now)
            return;
        membersInCooldown.delete(key);
    });
}, 30000);
// deno-lint-ignore require-await
exports.inhibitors.set("ownerOnly", async (bot, command, options) => {
    if (command.ownerOnly &&
        (!options?.author.id || !bot.owners?.includes(options.author.id)))
        return { type: errors_js_1.ErrorEnums.OWNER_ONLY };
    return true;
});
exports.inhibitors.set("botPermissions", async (bot, cmd, options) => {
    const command = cmd;
    if (command.botGuildPermissions?.length &&
        (!options?.guildId ||
            (0, mod_js_1.getMissingGuildPermissions)(bot, (await bot.cache.guilds.get(options.guildId)), (await bot.cache.members.get(bot.id, options.guildId)), command.botGuildPermissions).length))
        return {
            type: errors_js_1.ErrorEnums.BOT_MISSING_PERMISSIONS,
            channel: false,
            value: (0, mod_js_1.getMissingGuildPermissions)(bot, options.guild, (await bot.cache.members.get(bot.id, options?.guildId)), command.botGuildPermissions),
        };
    if (command.botChannelPermissions?.length &&
        (!options?.channel.id ||
            (await (0, mod_js_1.getMissingChannelPermissions)(bot, options.channel.id, bot.id, command.botChannelPermissions)).length))
        return {
            type: errors_js_1.ErrorEnums.BOT_MISSING_PERMISSIONS,
            channel: true,
            value: await (0, mod_js_1.getMissingChannelPermissions)(bot, options.channel.id, bot.id, command.botChannelPermissions),
        };
    return true;
});
exports.inhibitors.set("userPermissions", async (bot, cmd, options) => {
    const command = cmd;
    if (command.userGuildPermissions?.length &&
        (!options?.guildId ||
            !options.author.id ||
            (0, mod_js_1.getMissingGuildPermissions)(bot, (await bot.cache.guilds.get(options.guildId)), options.member, command.userGuildPermissions).length))
        return {
            type: errors_js_1.ErrorEnums.USER_MISSING_PERMISSIONS,
            channel: false,
            value: (0, mod_js_1.getMissingGuildPermissions)(bot, options.guild, options.member, command.userGuildPermissions),
        };
    if (command.userChannelPermissions?.length &&
        (!options?.guildId ||
            !options.author.id ||
            (await (0, mod_js_1.getMissingChannelPermissions)(bot, options.channel, options.member, command.userChannelPermissions)).length))
        return {
            type: errors_js_1.ErrorEnums.USER_MISSING_PERMISSIONS,
            channel: true,
            value: (0, mod_js_1.getMissingGuildPermissions)(bot, options.guild, options.member, command.userChannelPermissions),
        };
    return true;
});
