// discordeno@18.0.1/esm/rest/runMethod.js
import * as dntShim from "../_dnt.shims.js";
import { API_VERSION, BASE_URL, baseEndpoints } from "../util/constants.js";
import { encode } from "../util/base64.js";
export async function runMethod(rest, method, route, body, options) {
    rest.debug(`[REST - RequestCreate] Method: ${method} | URL: ${route} | Retry Count: ${options?.retryCount ?? 0} | Bucket ID: ${options?.bucketId} | Body: ${JSON.stringify(body)}`);
    const errorStack = new Error("Location:");
    // @ts-ignore Breaks deno deploy. Luca said add ts-ignore until it's fixed
    Error.captureStackTrace(errorStack);
    // For proxies we don't need to do any of the legwork so we just forward the request
    if (!baseEndpoints.BASE_URL.startsWith(BASE_URL) && route[0] === "/") {
        // Special handling for sending blobs across http to proxy
        if (body?.file) {
            if (!Array.isArray(body.file)) {
                body.file = [body.file];
            }
            // convert blobs to string before sending to proxy
            body.file = await Promise.all(body.file.map(async (f) => {
                const url = encode(await (f.blob).arrayBuffer());
                return { name: f.name, blob: `data:${f.blob.type};base64,${url}` };
            }));
        }
        const headers = {
            Authorization: rest.secretKey,
        };
        if (body) {
            headers["Content-Type"] = "application/json";
        }
        const result = await dntShim.fetch(`${baseEndpoints.BASE_URL}${route}`, {
            body: body ? JSON.stringify(body) : undefined,
            headers,
            method,
        });
        if (!result.ok) {
            const err = await result.json().catch(() => { });
            // Legacy Handling to not break old code or when body is missing
            if (!err?.body)
                throw new Error(`Error: ${err.message ?? result.statusText}`);
            throw rest.convertRestError(errorStack, err);
        }
        return result.status !== 204 ? await result.json() : undefined;
    }
    // No proxy so we need to handle all rate limiting and such
    return new Promise((resolve, reject) => {
        rest.processRequest(rest, {
            url: route[0] === "/" ? `${BASE_URL}/v${API_VERSION}${route}` : route,
            method,
            reject: (data) => {
                const restError = rest.convertRestError(errorStack, data);
                reject(restError);
            },
            respond: (data) => resolve(data.status !== 204 ? JSON.parse(data.body ?? "{}") : undefined),
        }, {
            bucketId: options?.bucketId,
            body: body,
            retryCount: options?.retryCount ?? 0,
            headers: options?.headers,
        });
    });
}
