// discordeno/script/helpers/channels/threads/startThreadWithoutMessage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.startThreadWithoutMessage = void 0;
/**
 * Creates a thread without using a message as the thread's point of origin.
 *
 * @param bot - The bot instance to use to make the request.
 * @param channelId - The ID of the channel in which to create the thread.
 * @param options - The parameters to use for the creation of the thread.
 * @returns An instance of the created {@link Channel | Thread}.
 *
 * @remarks
 * Creating a private thread requires the server to be boosted.
 *
 * Fires a _Thread Create_ gateway event.
 *
 * @see {@link https://discord.com/developers/docs/resources/channel#start-thread-without-message}
 */
async function startThreadWithoutMessage(bot, channelId, options) {
    const result = await bot.rest.runMethod(bot.rest, "POST", bot.constants.routes.THREAD_START_PRIVATE(channelId), {
        name: options.name,
        auto_archive_duration: options.autoArchiveDuration,
        rate_limit_per_user: options.rateLimitPerUser,
        type: options.type,
        invitable: options.invitable,
        reason: options.reason,
    });
    return bot.transformers.channel(bot, {
        channel: result,
        guildId: result.guild_id ? bot.transformers.snowflake(result.guild_id) : undefined,
    });
}
exports.startThreadWithoutMessage = startThreadWithoutMessage;
