// discordeno/script/helpers/guilds/voice/editVoiceState.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.editUserVoiceState = exports.editOwnVoiceState = exports.updateBotVoiceState = void 0;
exports.updateBotVoiceState = editOwnVoiceState;
/**
 * Edits the voice state of the bot user.
 *
 * @param bot - The bot instance to use to make the request.
 * @param guildId - The ID of the guild in which to edit the voice state of the bot user.
 * @param options - The parameters for the edit of the voice state.
 *
 * @remarks
 * The {@link EditOwnVoiceState.channelId | channelId} property of the {@link options} object parameter must point to a stage channel, and the bot user must already have joined it.
 *
 * If attempting to unmute oneself:
 * - Requires the `MUTE_MEMBERS` permission.
 *
 * If attempting to request to speak:
 * - Requires the `REQUEST_TO_SPEAK` permission.
 *
 * @see {@link https://discord.com/developers/docs/resources/guild#modify-current-user-voice-state}
 */
async function editOwnVoiceState(bot, guildId, options) {
    return await bot.rest.runMethod(bot.rest, "PATCH", bot.constants.routes.UPDATE_VOICE_STATE(guildId), {
        channel_id: options.channelId,
        suppress: options.suppress,
        request_to_speak_timestamp: options.requestToSpeakTimestamp
            ? new Date(options.requestToSpeakTimestamp).toISOString()
            : options.requestToSpeakTimestamp,
    });
}
exports.editOwnVoiceState = editOwnVoiceState;
// TODO: Make the `userId` property of `options` its own parameter.
/**
 * Edits the voice state of another user.
 *
 * @param bot - The bot instance to use to make the request.
 * @param guildId - The ID of the guild in which to edit the voice state of the bot user.
 * @param options - The parameters for the edit of the voice state.
 *
 * @remarks
 * The {@link EditOwnVoiceState.channelId | channelId} property of the {@link options} object parameter must point to a stage channel, and the user must already have joined it.
 *
 * Requires the `MUTE_MEMBERS` permission.
 *
 * @see {@link https://discord.com/developers/docs/resources/guild#modify-current-user-voice-state}
 */
async function editUserVoiceState(bot, guildId, options) {
    return await bot.rest.runMethod(bot.rest, "PATCH", bot.constants.routes.UPDATE_VOICE_STATE(guildId, options.userId), {
        channel_id: options.channelId,
        suppress: options.suppress,
        user_id: options.userId,
    });
}
exports.editUserVoiceState = editUserVoiceState;
