// discordeno/script/helpers/guilds/automod/createAutomodRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createAutomodRule = void 0;
/**
 * Creates an automod rule in a guild.
 *
 * @param bot - The bot instance to use to make the request.
 * @param guildId - The ID of the guild to create the rule in.
 * @param options - The parameters for the creation of the rule.
 * @returns An instance of the created {@link AutoModerationRule}.
 *
 * @remarks
 * Requires the `MANAGE_GUILD` permission.
 *
 * Fires an _Auto Moderation Rule Create_ gateway event.
 *
 * @see {@link https://discord.com/developers/docs/resources/auto-moderation#create-auto-moderation-rule}
 */
async function createAutomodRule(bot, guildId, options) {
    const result = await bot.rest.runMethod(bot.rest, "POST", bot.constants.routes.AUTOMOD_RULES(guildId), {
        name: options.name,
        event_type: options.eventType,
        trigger_type: options.triggerType,
        trigger_metadata: {
            keyword_filter: options.triggerMetadata.keywordFilter,
            presets: options.triggerMetadata.presets,
            allow_list: options.triggerMetadata.allowList,
            mention_total_limit: options.triggerMetadata.mentionTotalLimit,
        },
        actions: options.actions.map((action) => ({
            type: action.type,
            metadata: action.metadata
                ? {
                    channel_id: action.metadata.channelId?.toString(),
                    duration_seconds: action.metadata.durationSeconds,
                }
                : undefined,
        })),
        enabled: options.enabled ?? true,
        exempt_roles: options.exemptRoles?.map((id) => id.toString()),
        exempt_channels: options.exemptChannels?.map((id) => id.toString()),
        reason: options.reason,
    });
    return bot.transformers.automodRule(bot, result);
}
exports.createAutomodRule = createAutomodRule;
