// discordeno@18.0.1/script/rest/createRequestBody.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRequestBody = void 0;
const dntShim = __importStar(require("../_dnt.shims.js"));
const base64_js_1 = require("../util/base64.js");
const constants_js_1 = require("../util/constants.js");
/** Creates the request body and headers that are necessary to send a request. Will handle different types of methods and everything necessary for discord. */
function createRequestBody(rest, options) {
    const headers = {
        "user-agent": constants_js_1.USER_AGENT,
    };
    if (!options.unauthorized)
        headers["authorization"] = `Bot ${rest.token}`;
    // SOMETIMES SPECIAL HEADERS (E.G. CUSTOM AUTHORIZATION) NEED TO BE USED
    if (options.headers) {
        for (const key in options.headers) {
            headers[key.toLowerCase()] = options.headers[key];
        }
    }
    // GET METHODS SHOULD NOT HAVE A BODY
    if (options.method === "GET") {
        options.body = undefined;
    }
    // IF A REASON IS PROVIDED ENCODE IT IN HEADERS
    if (options.body?.reason) {
        headers["X-Audit-Log-Reason"] = encodeURIComponent(options.body.reason);
        options.body.reason = undefined;
    }
    // IF A FILE/ATTACHMENT IS PRESENT WE NEED SPECIAL HANDLING
    if (options.body?.file) {
        const files = findFiles(options.body.file);
        const form = new dntShim.FormData();
        // WHEN CREATING A STICKER, DISCORD WANTS FORM DATA ONLY
        if (options.url?.endsWith("/stickers") && options.method === "POST") {
            form.append(`file`, files[0].blob, files[0].name);
            form.append(`name`, options.body.name);
            form.append(`description`, options.body.description);
            form.append(`tags`, options.body.tags);
        }
        else {
            for (let i = 0; i < files.length; i++) {
                form.append(`file${i}`, files[i].blob, files[i].name);
            }
            form.append("payload_json", JSON.stringify({ ...options.body, file: undefined }));
        }
        options.body.file = form;
    }
    else if (options.body && !["GET", "DELETE"].includes(options.method)) {
        headers["Content-Type"] = "application/json";
    }
    return {
        headers,
        body: (options.body?.file ?? JSON.stringify(options.body)),
        method: options.method,
    };
}
exports.createRequestBody = createRequestBody;
function findFiles(file) {
    if (!file) {
        return [];
    }
    const files = Array.isArray(file) ? file : [file];
    return files.filter(coerceToFileContent);
}
function coerceToFileContent(value) {
    if (!value || typeof value !== "object") {
        return false;
    }
    const file = value;
    if (typeof file.name !== "string") {
        return false;
    }
    switch (typeof file.blob) {
        case "string": {
            const match = file.blob.match(/^data:(?<mimeType>[a-zA-Z0-9\/]*);base64,(?<content>.*)$/);
            if (match?.groups === undefined) {
                return false;
            }
            const { mimeType, content } = match.groups;
            file.blob = new dntShim.Blob([(0, base64_js_1.decode)(content)], { type: mimeType });
            return true;
        }
        case "object":
            return file.blob instanceof dntShim.Blob;
        default:
            return false;
    }
}
