// discordeno@18.0.1/script/rest/runMethod.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.runMethod = void 0;
const dntShim = __importStar(require("../_dnt.shims.js"));
const constants_js_1 = require("../util/constants.js");
const base64_js_1 = require("../util/base64.js");
async function runMethod(rest, method, route, body, options) {
    rest.debug(`[REST - RequestCreate] Method: ${method} | URL: ${route} | Retry Count: ${options?.retryCount ?? 0} | Bucket ID: ${options?.bucketId} | Body: ${JSON.stringify(body)}`);
    const errorStack = new Error("Location:");
    // @ts-ignore Breaks deno deploy. Luca said add ts-ignore until it's fixed
    Error.captureStackTrace(errorStack);
    // For proxies we don't need to do any of the legwork so we just forward the request
    if (!constants_js_1.baseEndpoints.BASE_URL.startsWith(constants_js_1.BASE_URL) && route[0] === "/") {
        // Special handling for sending blobs across http to proxy
        if (body?.file) {
            if (!Array.isArray(body.file)) {
                body.file = [body.file];
            }
            // convert blobs to string before sending to proxy
            body.file = await Promise.all(body.file.map(async (f) => {
                const url = (0, base64_js_1.encode)(await (f.blob).arrayBuffer());
                return { name: f.name, blob: `data:${f.blob.type};base64,${url}` };
            }));
        }
        const headers = {
            Authorization: rest.secretKey,
        };
        if (body) {
            headers["Content-Type"] = "application/json";
        }
        const result = await dntShim.fetch(`${constants_js_1.baseEndpoints.BASE_URL}${route}`, {
            body: body ? JSON.stringify(body) : undefined,
            headers,
            method,
        });
        if (!result.ok) {
            const err = await result.json().catch(() => { });
            // Legacy Handling to not break old code or when body is missing
            if (!err?.body)
                throw new Error(`Error: ${err.message ?? result.statusText}`);
            throw rest.convertRestError(errorStack, err);
        }
        return result.status !== 204 ? await result.json() : undefined;
    }
    // No proxy so we need to handle all rate limiting and such
    return new Promise((resolve, reject) => {
        rest.processRequest(rest, {
            url: route[0] === "/" ? `${constants_js_1.BASE_URL}/v${constants_js_1.API_VERSION}${route}` : route,
            method,
            reject: (data) => {
                const restError = rest.convertRestError(errorStack, data);
                reject(restError);
            },
            respond: (data) => resolve(data.status !== 204 ? JSON.parse(data.body ?? "{}") : undefined),
        }, {
            bucketId: options?.bucketId,
            body: body,
            retryCount: options?.retryCount ?? 0,
            headers: options?.headers,
        });
    });
}
exports.runMethod = runMethod;
