// discordeno@18.0.1/esm/helpers/roles/createRole.js
/**
 * Creates a role in a guild.
 *
 * @param bot - The bot instance to use to make the request.
 * @param guildId - The ID of the guild to create the role in.
 * @param options - The parameters for the creation of the role.
 * @returns An instance of the created {@link Role}.
 *
 * @remarks
 * Requires the `MANAGE_ROLES` permission.
 *
 * Fires a _Guild Role Create_ gateway event.
 *
 * @see {@link https://discord.com/developers/docs/resources/guild#create-guild-role}
 */
export async function createRole(bot, guildId, options, reason) {
    const result = await bot.rest.runMethod(bot.rest, "POST", bot.constants.routes.GUILD_ROLES(guildId), {
        name: options.name,
        color: options.color,
        hoist: options.hoist,
        mentionable: options.mentionable,
        permissions: bot.utils.calculateBits(options?.permissions || []),
        icon: options.icon,
        unicode_emoji: options.unicodeEmoji,
        reason,
    });
    return bot.transformers.role(bot, {
        role: result,
        guildId: bot.transformers.snowflake(guildId),
    });
}
