// discordeno/script/helpers/guilds/invites/getInvite.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getInvite = void 0;
const mod_js_1 = require("../../../mod.js");
/**
 * Gets an invite to a channel by its invite code.
 *
 * @param bot - The bot instance to use to make the request.
 * @param inviteCode - The invite code of the invite to get.
 * @param options - The parameters for the fetching of the invite.
 * @returns An instance of {@link BaseInvite | Invite}.
 *
 * @see {@link https://discord.com/developers/docs/resources/invite#get-invite}
 */
async function getInvite(bot, inviteCode, options) {
    const result = await bot.rest.runMethod(bot.rest, "GET", bot.constants.routes.INVITE(inviteCode, options));
    return {
        code: result.code,
        guildId: result.guild?.id ? bot.transformers.snowflake(result.guild.id) : undefined,
        channelId: result.channel?.id ? bot.transformers.snowflake(result.channel.id) : undefined,
        inviter: result.inviter ? bot.transformers.user(bot, result.inviter) : undefined,
        targetType: result.target_type
            ? (result.target_type === 1 ? mod_js_1.TargetTypes.Stream : mod_js_1.TargetTypes.EmbeddedApplication)
            : undefined,
        targetUser: result.target_user ? bot.transformers.user(bot, result.target_user) : undefined,
        targetApplicationId: result.target_application?.id
            ? bot.transformers.snowflake(result.target_application.id)
            : undefined,
        approximatePresenceCount: result.approximate_presence_count,
        approximateMemberCount: result.approximate_member_count,
        expiresAt: result.expires_at ? Date.parse(result.expires_at) : undefined,
    };
}
exports.getInvite = getInvite;
