// discordeno@18.0.1/esm/helpers/guilds/createGuild.js
/**
 * Creates a guild.
 *
 * @param bot - The bot instance to use to make the request.
 * @param options - The parameters for the creation of the guild.
 * @returns An instance of the created {@link Guild}.
 *
 * @remarks
 * ⚠️ This route can only be used by bots in __fewer than 10 guilds__.
 *
 * Fires a _Guild Create_ gateway event.
 *
 * @see {@link https://discord.com/developers/docs/resources/guild#create-guild}
 */
export async function createGuild(bot, options) {
    const result = await bot.rest.runMethod(bot.rest, "POST", bot.constants.routes.GUILDS(), {
        name: options.name,
        afk_channel_id: options.afkChannelId,
        afk_timeout: options.afkTimeout,
        channels: options.channels,
        default_message_notifications: options.defaultMessageNotifications,
        explicit_content_filter: options.explicitContentFilter,
        icon: options.icon,
        roles: options.roles,
        system_channel_flags: options.systemChannelFlags,
        system_channel_id: options.systemChannelId,
        verification_level: options.verificationLevel,
    });
    return bot.transformers.guild(bot, { guild: result, shardId: 0 });
}
