// discordeno/esm/helpers/guilds/voice/connectToVoiceChannel.js
import { GatewayOpcodes } from "../../../types/shared.js";
/**
 * Connects the bot user to a voice or stage channel.
 *
 * This function sends the _Update Voice State_ gateway command over the gateway behind the scenes.
 *
 * @param bot - The bot instance to use to make the request.
 * @param guildId - The ID of the guild the voice channel to leave is in.
 *
 * @remarks
 * Requires the `CONNECT` permission.
 *
 * Fires a _Voice State Update_ gateway event.
 *
 * @see {@link https://discord.com/developers/docs/topics/gateway#update-voice-state}
 */
export async function connectToVoiceChannel(bot, guildId, channelId, options) {
    const shardId = bot.utils.calculateShardId(bot.gateway, bot.transformers.snowflake(guildId));
    const shard = bot.gateway.manager.shards.get(shardId);
    if (!shard) {
        throw new Error(`Shard (id: ${shardId} not found`);
    }
    return shard.send({
        op: GatewayOpcodes.VoiceStateUpdate,
        d: {
            guild_id: guildId.toString(),
            channel_id: channelId.toString(),
            self_mute: Boolean(options?.selfMute),
            self_deaf: options?.selfDeaf ?? true,
        },
    });
}
