// discordeno/script/helpers/roles/editRole.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.editRole = void 0;
/**
 * Edits a role in a guild.
 *
 * @param bot - The bot instance to use to make the request.
 * @param guildId - The ID of the guild to edit the role in.
 * @param roleId - The ID of the role to edit.
 * @param options - The parameters for the edit of the role.
 * @returns An instance of the edited {@link Role}.
 *
 * @remarks
 * Requires the `MANAGE_ROLES` permission.
 *
 * Fires a _Guild Role Update_ gateway event.
 *
 * @see {@link https://discord.com/developers/docs/resources/guild#modify-guild-role}
 */
async function editRole(bot, guildId, roleId, options) {
    const result = await bot.rest.runMethod(bot.rest, "PATCH", bot.constants.routes.GUILD_ROLE(guildId, roleId), {
        name: options.name,
        color: options.color,
        hoist: options.hoist,
        mentionable: options.mentionable,
        permissions: bot.utils.calculateBits(options?.permissions || []),
        icon: options.icon,
        unicode_emoji: options.unicodeEmoji,
    });
    return bot.transformers.role(bot, { role: result, guildId: bot.transformers.snowflake(guildId) });
}
exports.editRole = editRole;
